processing/src/main/java/io/druid/query/extraction/namespace/ExtractionNamespaceFunctionFactory.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.extraction.namespace;
+
+import com.google.common.base.Function;
+
+import java.util.Map;
+
+/**
+ *
+ */
+public interface ExtractionNamespaceFunctionFactory<T extends ExtractionNamespace>
+{
+  Function<String, String> build(T extractionNamespace);
+
+  Runnable getCachePopulator(T extractionNamespace, Map<String, String> swap);
```

Comment:
"runnable" = "cache populator runnable" for the comment below

I see that first time when runnable runs the "swap" will be available for read only after runnable is done with it. However, on second(and subsequent) runnable runs it seems still possible that same cache map is being read for some query and getting updated by the runnable at the same time. So "live" copy of cached data can still be written while being read.... no?